### PR TITLE
feat: add 'gnokey maketx run'

### PR DESCRIPTION
**Pull request by moul**
With the `gnokey maketx run` command, you can execute any code directly on the chain from your user realm. It's convenient because it doesn't require uploading a whole new package - it simply creates a temporary `memfile`.

After merging this pull request, our next goal is to develop a client that enables a chain-backed REPL, providing a shell experience similar to Unix. This client will utilize `q_eval` for reading and `run` for writing, enabling seamless interaction with the chain directly from a Go shell. However, for now, let's focus on the current plan.

-  [x] Update CLI
-  [x] Implement RPC handler
-  [x] Perform unit and integration tests
-  [x] Restrict to `package main`
-  [x] Fix: `tests.GetPrevRealm(struct{("g1vp3uusf5txt06lx76x7qq70nldh5n2hcjqcs88" std.Address),("gno.land/r/g123456/run", string)} std.Realm)` -> finally won't fix it; we want it to be seen as "gno.land/r/{caller.Addr()}/run"
-  [x] Use `main` instead of `Main`
-  [x] Blacklist `addpkg -pkgpath gno.land/r/main` (reserved)

Bonuses and notes for later -> https://github.com/gnolang/gno/issues/1283

### CODE REVIEW

**Comment by moul** [comment]
```console
$ cat /tmp/script.gno
package main

import (
    "std"

    "gno.land/r/demo/tests"
    "gno.land/r/demo/users"
)

func Main() {
    println("HELLO WORLD3x!")
    println("render", users.Render(""))
    println("tests.CurrentRealmPath", tests.CurrentRealmPath())
    println("tests.IsOriginCall", tests.IsOriginCall())
    println("tests.GetPrevRealm", tests.GetPrevRealm())
    println("std.GetOrigCaller", std.GetOrigCaller())
    println("std.PrevRealm", std.PrevRealm())
}
$ go run ./gno.land/cmd/gnokey maketx run moul -gas-wanted 10000000 -gas-fee 10000000ugnot -broadcast /tmp/script.gno
Enter password.
HELLO WORLD3x!
render  * [administrator](/r/demo/users:administrator)
 * [anarcher](/r/demo/users:anarcher)
 * [ideamour](/r/demo/users:ideamour)
 * [manfred](/r/demo/users:manfred)
 * [piupiu](/r/demo/users:piupiu)
 * [zo_oma](/r/demo/users:zo_oma)

tests.CurrentRealmPath gno.land/r/demo/tests
tests.IsOriginCall false
tests.GetPrevRealm (struct{("g1vp3uusf5txt06lx76x7qq70nldh5n2hcjqcs88" std.Address),("gno.land/r/main" string)} std.Realm)
std.GetOrigCaller g1u7y667z64x2h7vc6fmpcprgey4ck233jaww9zq
std.PrevRealm (struct{("g1u7y667z64x2h7vc6fmpcprgey4ck233jaww9zq" std.Address),("" string)} std.Realm)

OK!
GAS WANTED: 10000000
GAS USED:   401843
```

**Comment by moul** [comment]
New demo:

CleanShot 2023-10-19 at 09 05 03@2x

File: tm2/pkg/crypto/keys/client/run.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+func newRunCmd(rootCfg *makeTxCfg, io *commands.IO) *commands.Command {
+	cfg := &runCfg{
+		rootCfg: rootCfg,
+	}
+
+	return commands.NewCommand(
+		commands.Metadata{
+			Name:       "run",
+			ShortUsage: "run [flags] <key-name or address> <file or - or dir>",
+			ShortHelp:  "utes arbitrary Gno code",
```
[the remark this answers is not in the document]
Comment (moul) [reply tm2/pkg/crypto/keys/client/run.go:33]:
```suggestion
			ShortHelp:  "Executes arbitrary Gno code",
```

File: gno.land/pkg/sdk/vm/keeper.go [hunk 2]
```diff
@@ -282,6 +284,89 @@ func (vm *VMKeeper) Call(ctx sdk.Context, msg MsgCall) (res string, err error) {
...
+		OrigSend:      send,
+		OrigSendSpent: new(std.Coins),
+		OrigPkgAddr:   pkgAddr.Bech32(),
+		Banker:        NewSDKBanker(vm, ctx),
+	}
+	// Parse and run the files, construct *PV.
+	buf := new(bytes.Buffer)
+	m := gno.NewMachineWithOptions(
+		gno.MachineOptions{
+			PkgPath:   "",
```
Comment (jaekwon) [inline gno.land/pkg/sdk/vm/keeper.go:326]: shouldn't this be "main"?

**Review by jaekwon — APPROVED** [review]
i didn't review every line but generally looks great!

File: gno.land/pkg/sdk/vm/keeper.go (hunk 2 above)
Comment (moul) [reply gno.land/pkg/sdk/vm/keeper.go:336]: to be honest, I'm really not sure.

**Review by thehowl — COMMENTED** [review]
So, I started reviewing individual files and added a few comments but didn't finish.

Essentially, my take on this is that we should have only one of run or call.

Call could be an alias for run, so that what is implicitly sent to the chain is `package main; func main() { FunctionCall(args...) }`.

Wdyt?

File: tm2/pkg/crypto/keys/client/run.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+func newRunCmd(rootCfg *makeTxCfg, io *commands.IO) *commands.Command {
+	cfg := &runCfg{
+		rootCfg: rootCfg,
+	}
+
+	return commands.NewCommand(
+		commands.Metadata{
+			Name:       "run",
+			ShortUsage: "run [flags] <key-name or address> <file or - or dir>",
```
[the remark this answers is not in the document]
Comment (moul) [reply tm2/pkg/crypto/keys/client/run.go:32]: If you put a dir, it should load a mempkg from the dir, validate, and then try to call the main function.

File: tm2/pkg/crypto/keys/client/run.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+			ShortHelp:  "Executes arbitrary Gno code",
```
[the remark this answers is not in the document]
Comment (moul) [reply tm2/pkg/crypto/keys/client/run.go:33]: Oh yes!

**Comment by moul** [comment]
> Essentially, my take on this is that we should have only one of run or call.
> 
> Call could be an alias for run, so that what is implicitly sent to the chain is `package main; func main() { FunctionCall(args...) }`.

I had the same feeling. However, the 'call' function is efficient and should be retained due to its minimal parsing overhead. I'm considering adding a dry-run or readonly mode without transactions on “run”, possibly as an optional feature for local full node testing, not for public use.

**Comment by thehowl** [comment]
The summary of how I think we should be going:

- `qeval` and `msg_call` should have the same basic input: an expression which may return a value which is evaluated. With qeval, this is just a query and cannot modify state, while msg_call creates a tx.
- `msg_call` can, additionally, be enriched with some extra code, which is parsed as a normal Gno file.
  - This code is executed in the "realm" of the user; any contract calls from that code will have PrevRealm == OrigCaller.
  - msg_call's main parameter may call the functions defined in this realm. Ie. `--address my_address --expr 'main()'` will be essentially the way that `maketx run` is working now.

**Pull request merged**